{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# This notebook provides an example script for running association test\n",
    "1. Install all required packages\n",
    "2. Download required files from the `files_for_association` folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import math\n",
    "from tqdm import tqdm\n",
    "import random\n",
    "import statsmodels.api as sm\n",
    "from sklearn import preprocessing\n",
    "import statsmodels.stats.multitest\n",
    "\n",
    "def getData(filename):\n",
    "    with open(filename, \"r\") as csvfile:\n",
    "        reader = csv.reader(csvfile)\n",
    "        for row in reader:\n",
    "            yield row\n",
    "\n",
    "def exonstrSLR(chrom,Pop,files_dir,minTs=10,mingt=3,minsPgt=3,search_range=100_000):\n",
    "    \"\"\"\n",
    "    Pop: population \n",
    "    minTs: minimum samples required for regression\n",
    "    mingt: minimum genotypes required for regression\n",
    "    minsPgt: minimum samples requried for genotypes to be included in regression\n",
    "    search_range: ranges search for associated STRs\n",
    "    \"\"\"\n",
    "\n",
    "    #load covariates\n",
    "    cov_df=pd.read_csv(f'{files_dir}/covariates_all.csv',sep='\\t',index_col='sample_id')\n",
    "    exp_df=pd.read_csv(f'{files_dir}/'+Pop+'_normalized_and_filtered_hg38_chr'+str(chrom)+'.csv',sep='\\t')\n",
    "    gt_dir=f'{files_dir}/chr'+str(chrom)+'.GB.FORMAT'\n",
    "    \n",
    "    csv_gen = getData(gt_dir)\n",
    "    row_count = 0\n",
    "    p_df = pd.DataFrame(columns =[ \"str-gene\",\"str_end\",\"motif\",\"gene_name\", \"sample_n\",\"GT_n\",\"p_values\",\"slope\",'error',\"shuffled_p\",\"shuffled_slope\",\"shuffled_error\"])\n",
    "    #the total length just for illustration purpose, can remove tqdm when running in large batches\n",
    "    for row in tqdm(csv_gen,total=1067):\n",
    "        if row_count==0:\n",
    "            #extract sample name of current STR\n",
    "            col_name=row[0].split('\\t')\n",
    "            row_count+=1\n",
    "            continue\n",
    "\n",
    "        gt_value=list(filter(lambda a: a!='',row[0].split('\\t')))\n",
    "        gt_df=pd.DataFrame([gt_value],columns=col_name)\n",
    "        gt_df[gt_df.columns[4:]]=gt_df[gt_df.columns[4:]]. \\\n",
    "                applymap(lambda x: [int(x.split('/')[0]),int(x.split('/')[1])] if '/' in x else [None,None])\n",
    "        gt_samples=gt_df.columns[4:][gt_df.iloc[0,4:].apply(lambda x: x !=[None,None])].to_list()\n",
    "        gt_psi_samples=list(set(exp_df.columns[5:]) & set(gt_samples))\n",
    "        #STR GENOTYPE FILTER\n",
    "        gt_psi_phased=gt_df[gt_psi_samples].T.rename(columns={0:'GT'}).applymap(lambda x: sum(x))\n",
    "        gt_sum=gt_psi_phased.groupby('GT').size()\n",
    "\n",
    "        #get joint data\n",
    "        gt_ab_3=gt_sum[gt_sum>=minsPgt].index.tolist()\n",
    "        #check number of genotype and check how many samples remains\n",
    "        if len(gt_ab_3)<mingt or gt_sum[gt_ab_3].sum()<minTs:\n",
    "            row_count+=1\n",
    "            continue\n",
    "\n",
    "        gt_psi_filted_samples=set(gt_psi_phased[gt_psi_phased.GT.isin(gt_ab_3)].index) & set(cov_df.index)\n",
    "        gt_psi_samples=list(gt_psi_filted_samples)\n",
    "\n",
    "        gt_df=gt_df[gt_df.columns[0:4].to_list()+gt_psi_samples]\n",
    "        joint_cov_df=cov_df.loc[gt_psi_samples]\n",
    "        joint_exp_df=exp_df[exp_df.columns[0:5].tolist()+gt_psi_samples]\n",
    "        joint_exp_df=joint_exp_df[joint_exp_df.chromosome.values == gt_df['CHROM'].values]\n",
    "\n",
    "        paired_df=joint_exp_df[(joint_exp_df.start - search_range <= int(gt_df.POS)) \\\n",
    "                         & (joint_exp_df.end + search_range >= (int(gt_df.POS)))]\n",
    "\n",
    "        if paired_df.empty:\n",
    "            row_count+=1\n",
    "            continue\n",
    "\n",
    "        for index, curr_exon in paired_df.iterrows():\n",
    "            #contat the PSI, genotype, peer and pc for current exon of all samples\n",
    "            a=pd.concat([curr_exon[5:].astype('float64'),\\\n",
    "                       gt_df[gt_psi_samples].T.rename(columns={0:'GT'}).applymap(lambda x: sum(x)),\\\n",
    "                       joint_cov_df],\\\n",
    "                       axis=1).rename(columns={index:'exp'})\n",
    "\n",
    "            a=a[~a.exp.isnull()]\n",
    "            #standardization\n",
    "            a_scaled = preprocessing.StandardScaler().fit_transform(a)\n",
    "            y=a_scaled[:,0]\n",
    "            x=a_scaled[:,1:]\n",
    "            x=sm.add_constant(x)\n",
    "            mod_ols  = sm.OLS(y,x)\n",
    "            res_ols = mod_ols.fit()\n",
    "            p_values=res_ols.pvalues[1]\n",
    "            slope=res_ols.params[1]\n",
    "            err=res_ols.bse[1]\n",
    "\n",
    "            shuffled_y = random.sample(list(y),len(y))\n",
    "            mod_ols_s = sm.OLS(shuffled_y,x)\n",
    "            res_ols_s = mod_ols_s.fit()\n",
    "            shuffled_p=res_ols_s.pvalues[1]\n",
    "            slope_p=res_ols_s.params[1]\n",
    "            err_p=res_ols_s.bse[1]\n",
    "\n",
    "            p_df = p_df.append({\"str-gene\":list(gt_df.CHROM +'_'+ gt_df.POS.str.rstrip()+'-'+curr_exon.gene_id)[0],\\\n",
    "                                \"str_end\":gt_df.END.tolist()[0],\\\n",
    "                                \"motif\":gt_df.motif.tolist()[0],\\\n",
    "                                \"gene_name\":curr_exon.gene_name,\"sample_n\":len(a),\"GT_n\":len(gt_ab_3),\"p_values\":p_values,\\\n",
    "                                \"slope\":slope,\"error\":err,\"shuffled_p\":shuffled_p,\"shuffled_slope\":slope_p,\\\n",
    "                                \"shuffled_error\":err_p}, ignore_index=True)\n",
    "            row_count+=1\n",
    "\n",
    "    return p_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 75%|███████▌  | 802/1067 [01:31<00:30,  8.75it/s]\n"
     ]
    }
   ],
   "source": [
    "#directory where association files are lcoated\n",
    "reg_file_dir=\"/expanse/protected/gymreklab-dbgap/mount/yal084/genotyping_repeats_tutorial/files_for_association\"\n",
    "#running regression\n",
    "reg_results=exonstrSLR(11,\"AFR\",reg_file_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>str-gene</th>\n",
       "      <th>str_end</th>\n",
       "      <th>motif</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>sample_n</th>\n",
       "      <th>GT_n</th>\n",
       "      <th>p_values</th>\n",
       "      <th>slope</th>\n",
       "      <th>error</th>\n",
       "      <th>shuffled_p</th>\n",
       "      <th>shuffled_slope</th>\n",
       "      <th>shuffled_error</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr11_57429188-ENSG00000134809.4</td>\n",
       "      <td>57429198</td>\n",
       "      <td>T</td>\n",
       "      <td>TIMM10</td>\n",
       "      <td>89</td>\n",
       "      <td>3</td>\n",
       "      <td>0.560010</td>\n",
       "      <td>0.060683</td>\n",
       "      <td>0.103031</td>\n",
       "      <td>0.368764</td>\n",
       "      <td>0.148819</td>\n",
       "      <td>0.163240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr11_57445499-ENSG00000134809.4</td>\n",
       "      <td>57445555</td>\n",
       "      <td>AAAAA</td>\n",
       "      <td>TIMM10</td>\n",
       "      <td>84</td>\n",
       "      <td>6</td>\n",
       "      <td>0.860272</td>\n",
       "      <td>0.022799</td>\n",
       "      <td>0.128291</td>\n",
       "      <td>0.132516</td>\n",
       "      <td>0.244131</td>\n",
       "      <td>0.157390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr11_57446656-ENSG00000134809.4</td>\n",
       "      <td>57446668</td>\n",
       "      <td>T</td>\n",
       "      <td>TIMM10</td>\n",
       "      <td>89</td>\n",
       "      <td>3</td>\n",
       "      <td>0.958425</td>\n",
       "      <td>-0.006655</td>\n",
       "      <td>0.126673</td>\n",
       "      <td>0.018731</td>\n",
       "      <td>-0.437541</td>\n",
       "      <td>0.176658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr11_57447780-ENSG00000134809.4</td>\n",
       "      <td>57447790</td>\n",
       "      <td>A</td>\n",
       "      <td>TIMM10</td>\n",
       "      <td>84</td>\n",
       "      <td>3</td>\n",
       "      <td>0.962715</td>\n",
       "      <td>-0.005570</td>\n",
       "      <td>0.118056</td>\n",
       "      <td>0.414625</td>\n",
       "      <td>-0.136463</td>\n",
       "      <td>0.164701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chr11_57448212-ENSG00000134809.4</td>\n",
       "      <td>57448235</td>\n",
       "      <td>AAGC</td>\n",
       "      <td>TIMM10</td>\n",
       "      <td>85</td>\n",
       "      <td>3</td>\n",
       "      <td>0.513918</td>\n",
       "      <td>-0.074486</td>\n",
       "      <td>0.112660</td>\n",
       "      <td>0.828003</td>\n",
       "      <td>-0.033264</td>\n",
       "      <td>0.151677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483</th>\n",
       "      <td>chr11_57617027-ENSG00000134809.4</td>\n",
       "      <td>57617039</td>\n",
       "      <td>A</td>\n",
       "      <td>TIMM10</td>\n",
       "      <td>89</td>\n",
       "      <td>3</td>\n",
       "      <td>0.302509</td>\n",
       "      <td>0.124972</td>\n",
       "      <td>0.119252</td>\n",
       "      <td>0.970076</td>\n",
       "      <td>0.006531</td>\n",
       "      <td>0.172752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>493</th>\n",
       "      <td>chr11_57619285-ENSG00000134809.4</td>\n",
       "      <td>57619317</td>\n",
       "      <td>GA</td>\n",
       "      <td>TIMM10</td>\n",
       "      <td>86</td>\n",
       "      <td>3</td>\n",
       "      <td>0.100952</td>\n",
       "      <td>0.194828</td>\n",
       "      <td>0.115000</td>\n",
       "      <td>0.995314</td>\n",
       "      <td>0.001078</td>\n",
       "      <td>0.181994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>503</th>\n",
       "      <td>chr11_57621715-ENSG00000134809.4</td>\n",
       "      <td>57621742</td>\n",
       "      <td>T</td>\n",
       "      <td>TIMM10</td>\n",
       "      <td>86</td>\n",
       "      <td>4</td>\n",
       "      <td>0.879389</td>\n",
       "      <td>-0.019005</td>\n",
       "      <td>0.124141</td>\n",
       "      <td>0.269169</td>\n",
       "      <td>-0.207219</td>\n",
       "      <td>0.183941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>513</th>\n",
       "      <td>chr11_57624531-ENSG00000134809.4</td>\n",
       "      <td>57624566</td>\n",
       "      <td>TATT</td>\n",
       "      <td>TIMM10</td>\n",
       "      <td>84</td>\n",
       "      <td>6</td>\n",
       "      <td>0.436481</td>\n",
       "      <td>0.091612</td>\n",
       "      <td>0.115983</td>\n",
       "      <td>0.999652</td>\n",
       "      <td>0.000064</td>\n",
       "      <td>0.145248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>523</th>\n",
       "      <td>chr11_57626417-ENSG00000134809.4</td>\n",
       "      <td>57626429</td>\n",
       "      <td>T</td>\n",
       "      <td>TIMM10</td>\n",
       "      <td>87</td>\n",
       "      <td>3</td>\n",
       "      <td>0.189979</td>\n",
       "      <td>0.140389</td>\n",
       "      <td>0.104688</td>\n",
       "      <td>0.486753</td>\n",
       "      <td>0.100833</td>\n",
       "      <td>0.143192</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             str-gene   str_end  motif gene_name sample_n  \\\n",
       "6    chr11_57429188-ENSG00000134809.4  57429198      T    TIMM10       89   \n",
       "12   chr11_57445499-ENSG00000134809.4  57445555  AAAAA    TIMM10       84   \n",
       "19   chr11_57446656-ENSG00000134809.4  57446668      T    TIMM10       89   \n",
       "26   chr11_57447780-ENSG00000134809.4  57447790      A    TIMM10       84   \n",
       "33   chr11_57448212-ENSG00000134809.4  57448235   AAGC    TIMM10       85   \n",
       "..                                ...       ...    ...       ...      ...   \n",
       "483  chr11_57617027-ENSG00000134809.4  57617039      A    TIMM10       89   \n",
       "493  chr11_57619285-ENSG00000134809.4  57619317     GA    TIMM10       86   \n",
       "503  chr11_57621715-ENSG00000134809.4  57621742      T    TIMM10       86   \n",
       "513  chr11_57624531-ENSG00000134809.4  57624566   TATT    TIMM10       84   \n",
       "523  chr11_57626417-ENSG00000134809.4  57626429      T    TIMM10       87   \n",
       "\n",
       "    GT_n  p_values     slope     error  shuffled_p  shuffled_slope  \\\n",
       "6      3  0.560010  0.060683  0.103031    0.368764        0.148819   \n",
       "12     6  0.860272  0.022799  0.128291    0.132516        0.244131   \n",
       "19     3  0.958425 -0.006655  0.126673    0.018731       -0.437541   \n",
       "26     3  0.962715 -0.005570  0.118056    0.414625       -0.136463   \n",
       "33     3  0.513918 -0.074486  0.112660    0.828003       -0.033264   \n",
       "..   ...       ...       ...       ...         ...             ...   \n",
       "483    3  0.302509  0.124972  0.119252    0.970076        0.006531   \n",
       "493    3  0.100952  0.194828  0.115000    0.995314        0.001078   \n",
       "503    4  0.879389 -0.019005  0.124141    0.269169       -0.207219   \n",
       "513    6  0.436481  0.091612  0.115983    0.999652        0.000064   \n",
       "523    3  0.189979  0.140389  0.104688    0.486753        0.100833   \n",
       "\n",
       "     shuffled_error  \n",
       "6          0.163240  \n",
       "12         0.157390  \n",
       "19         0.176658  \n",
       "26         0.164701  \n",
       "33         0.151677  \n",
       "..              ...  \n",
       "483        0.172752  \n",
       "493        0.181994  \n",
       "503        0.183941  \n",
       "513        0.145248  \n",
       "523        0.143192  \n",
       "\n",
       "[67 rows x 12 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reg_results[(reg_results.gene_name==\"TIMM10\") ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
